### PR TITLE
Check images before creating maia annotations

### LINKED ISSUE #121: Filter MAIA annotations before insert
https://github.com/biigle/maia/blob/e6bb111590ff55c95bb4401aab6948730e353b00/src/Jobs/JobResponse.php#L75-L88

There was an error where an image that MAIA generated annotations for was deleted in the meantime. The image IDs should be checked after `createMaiaAnnotation()` and before `insertAnnotationChunk()` and annotations of missing images should be removed. Also create a test case for this.

### PR DESCRIPTION
**Pull request by lehecht**
- [x] Check that the image exists before processing annotations
- [x] Add test case for deleted image

Closes #121.

### CODE REVIEW

File: src/Jobs/DetectionJob.php [hunk 1]
```diff
@@ -217,6 +219,14 @@ protected function createMaiaAnnotations(array $annotations)
             return $this->createMaiaAnnotation($annotation);
         }, $annotations);
 
+        // Remove annotations of already deleted images.        
+        foreach ($maiaAnnotations as $idx => $a) {
+            if (Image::where('id', '=', $a['image_id'])->doesntExist()) {
+                ImageAnnotation::where('image_id', '=', $a['image_id'])->delete();
```
Comment (mzur) [inline src/Jobs/DetectionJob.php:225]: Don't take my issue description literally. Or do you use ChatGPT? :smile: 

The actual image annotations should not be deleted, of course. What's more, if the image no longer exists, there are not annotations to delete anyway.

File: src/Jobs/DetectionJob.php [hunk 2]
```diff
@@ -217,6 +219,14 @@ protected function createMaiaAnnotations(array $annotations)
             return $this->createMaiaAnnotation($annotation);
         }, $annotations);
 
+        // Remove annotations of already deleted images.        
+        foreach ($maiaAnnotations as $idx => $a) {
+            if (Image::where('id', '=', $a['image_id'])->doesntExist()) {
```
Comment (mzur) [inline src/Jobs/DetectionJob.php:224]: This is really inefficient. You run one or two database queries for each of the possibly tens of thousands of MAIA annotations. Better collect the (distinct) image IDs first, then run a query (maybe chunk if there are too many images) that checks which of these IDs no longer exists. Finally `unset()` all MAIA annotations for images that no longer exist.

**Review by mzur — CHANGES_REQUESTED** [review]

File: src/Jobs/DetectionJob.php (hunk 1 above)
Comment (mzur) [reply src/Jobs/DetectionJob.php:225]: I meant `$maiaAnnotations` to be inserted into the DB should be removed from the array before the insert operation if the image of the MAIA annotation no longer exists. Your code already does this. But this has nothing to do with the actual `ImageAnnotation`.

Comment (lehecht) [reply src/Jobs/DetectionJob.php:225]: ah, okay. Now I understand. Thanks!

File: src/Jobs/DetectionJob.php [hunk 3]
```diff
@@ -217,6 +219,26 @@ protected function createMaiaAnnotations(array $annotations)
...
+                if($isDeleted) {
+                    $deletedImgId[$imgId] = true;
+                    return false;
+                } else {
+                    $deletedImgId[$imgId] = false;
+                }
+            }
+            return true;
+        });
+
```
Comment (mzur) [inline src/Jobs/DetectionJob.php:241]: This still performs one database query for each image. Also, can you modify `$deletedImgId` from within the closure at all? If you don't `use (&$deletedImgId)`, the array is passed by value, not by reference.

What I meant was something like this:
```php
$existingIds = $this->job->volume->images()->pluck('id', 'id');
$maiaAnnotations = array_filter($maiaAnnotations, fn ($a) => array_key_exists($a['image_id'], $existingIds));
```

This runs only a single query.

File: tests/Jobs/ObjectDetectionTest.php
Comment (mzur) [inline tests/Jobs/ObjectDetectionTest.php:1]: Can you please reduce the test assertions to only those that are relevant for this particular test case  (e.g. the JSON content checks are not relevant)? It's hard to find the relevant stuff among all the other assertions.

**Review by mzur — CHANGES_REQUESTED** [review]

**Review by mzur — APPROVED** [review]
Thanks! I've just simplified the test a little more.

**Pull request merged**